Exception when calling InstalledAppsData#queryInstalledApps

Hi to all, 

I am using OSHI to call InstalledAppsData#queryInstalledApps to check if a certain app is installed. While on other machines (e.g. windows server 2022) this is working fine, on a windows server 2019 it crahes with the following stack: 
Caused by: java.lang.RuntimeException: Unexpected registry type 4, expected REG_SZ or REG_EXPAND_SZ
	at com.sun.jna.platform.win32.Advapi32Util.registryGetStringValue(Advapi32Util.java:774)
	at com.sun.jna.platform.win32.Advapi32Util.registryGetStringValue(Advapi32Util.java:744)
	at oshi.driver.windows.registry.InstalledAppsData.getRegistryValueOrUnknown(InstalledAppsData.java:92)
	at oshi.driver.windows.registry.InstalledAppsData.queryInstalledApps(InstalledAppsData.java:63)
	at oshi.software.os.windows.WindowsInstalledApps.queryInstalledApps(WindowsInstalledApps.java:18)
	at oshi.util.Memoizer$1.get(Memoizer.java:65)
	at oshi.software.os.windows.WindowsOperatingSystem.getInstalledApplications(WindowsOperatingSystem.java:538)

Could you please make this more failsafe?

--- a/oshi-core/src/main/java/oshi/driver/windows/registry/InstalledAppsData.java
+++ b/oshi-core/src/main/java/oshi/driver/windows/registry/InstalledAppsData.java
@@ -4,81 +4,107 @@
  */
 package oshi.driver.windows.registry;
 
-import com.sun.jna.platform.win32.Advapi32Util;
-import com.sun.jna.platform.win32.Win32Exception;
-import com.sun.jna.platform.win32.WinNT;
-import com.sun.jna.platform.win32.WinReg;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-import oshi.software.os.ApplicationInfo;
-import oshi.util.ParseUtil;
+import static com.sun.jna.platform.win32.WinError.ERROR_INSUFFICIENT_BUFFER;
+import static com.sun.jna.platform.win32.WinError.ERROR_SUCCESS;
+import static com.sun.jna.platform.win32.WinNT.KEY_READ;
+import static com.sun.jna.platform.win32.WinNT.KEY_WOW64_32KEY;
+import static com.sun.jna.platform.win32.WinNT.KEY_WOW64_64KEY;
+import static com.sun.jna.platform.win32.WinNT.REG_DWORD;
+import static com.sun.jna.platform.win32.WinNT.REG_EXPAND_SZ;
+import static com.sun.jna.platform.win32.WinNT.REG_SZ;
+import static com.sun.jna.platform.win32.WinReg.HKEY_CURRENT_USER;
+import static com.sun.jna.platform.win32.WinReg.HKEY_LOCAL_MACHINE;
 
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.HashMap;
-import java.util.List;
 import java.util.LinkedHashMap;
 import java.util.LinkedHashSet;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
-import java.util.Arrays;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.sun.jna.Memory;
+import com.sun.jna.Native;
+import com.sun.jna.Pointer;
+import com.sun.jna.platform.win32.Advapi32;
+import com.sun.jna.platform.win32.Advapi32Util;
+import com.sun.jna.platform.win32.Win32Exception;
+import com.sun.jna.platform.win32.WinReg.HKEY;
+import com.sun.jna.platform.win32.WinReg.HKEYByReference;
+import com.sun.jna.ptr.IntByReference;
+import com.sun.jna.win32.W32APITypeMapper;
+
+import oshi.software.os.ApplicationInfo;
+import oshi.util.ParseUtil;
 
 public final class InstalledAppsData {
     private static final Logger LOG = LoggerFactory.getLogger(InstalledAppsData.class);
 
+    private static final Advapi32 ADV = Advapi32.INSTANCE;
+
+    private static final long THIRTY_YEARS_IN_SECS = 30L * 365 * 24 * 60 * 60;
+
     private InstalledAppsData() {
     }
 
-    private static final Map<WinReg.HKEY, List<String>> REGISTRY_PATHS = new HashMap<>();
-    private static final int[] ACCESS_FLAGS = { WinNT.KEY_WOW64_64KEY, WinNT.KEY_WOW64_32KEY };
+    private static final Map<HKEY, List<String>> REGISTRY_PATHS = new HashMap<>();
+    private static final int[] ACCESS_FLAGS = { KEY_WOW64_64KEY, KEY_WOW64_32KEY };
 
     static {
-        REGISTRY_PATHS.put(WinReg.HKEY_LOCAL_MACHINE,
-                Arrays.asList("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall",
-                        "SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\Uninstall"));
+        REGISTRY_PATHS.put(HKEY_LOCAL_MACHINE, Arrays.asList("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall",
+                "SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\Uninstall"));
 
-        REGISTRY_PATHS.put(WinReg.HKEY_CURRENT_USER,
-                Arrays.asList("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall"));
+        REGISTRY_PATHS.put(HKEY_CURRENT_USER, Arrays.asList("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall"));
     }
 
     public static List<ApplicationInfo> queryInstalledApps() {
         Set<ApplicationInfo> appInfoSet = new LinkedHashSet<>();
 
         // Iterate through both HKLM and HKCU paths
-        for (Map.Entry<WinReg.HKEY, List<String>> entry : REGISTRY_PATHS.entrySet()) {
-            WinReg.HKEY rootKey = entry.getKey();
+        for (Map.Entry<HKEY, List<String>> entry : REGISTRY_PATHS.entrySet()) {
+            HKEY rootKey = entry.getKey();
             List<String> uninstallPaths = entry.getValue();
 
             for (String registryPath : uninstallPaths) {
                 for (int accessFlag : ACCESS_FLAGS) {
-                    String[] keys = Advapi32Util.registryGetKeys(rootKey, registryPath, accessFlag);
-                    for (String key : keys) {
-                        String fullPath = registryPath + "\\" + key;
-                        try {
-                            String name = getRegistryValueOrUnknown(rootKey, fullPath, "DisplayName", accessFlag);
-                            if (name == null) {
-                                continue;
+                    try {
+                        String[] keys = Advapi32Util.registryGetKeys(rootKey, registryPath, accessFlag);
+                        for (String key : keys) {
+                            String fullPath = registryPath + "\\" + key;
+                            try {
+                                String name = registryValueToString(
+                                        getRegistryValueOrNull(rootKey, fullPath, "DisplayName", accessFlag));
+                                if (name == null) {
+                                    continue;
+                                }
+                                String version = registryValueToString(
+                                        getRegistryValueOrNull(rootKey, fullPath, "DisplayVersion", accessFlag));
+                                String publisher = registryValueToString(
+                                        getRegistryValueOrNull(rootKey, fullPath, "Publisher", accessFlag));
+                                long installDate = registryValueToLong(
+                                        getRegistryValueOrNull(rootKey, fullPath, "InstallDate", accessFlag));
+                                String installLocation = registryValueToString(
+                                        getRegistryValueOrNull(rootKey, fullPath, "InstallLocation", accessFlag));
+                                String installSource = registryValueToString(
+                                        getRegistryValueOrNull(rootKey, fullPath, "InstallSource", accessFlag));
+
+                                Map<String, String> additionalInfo = new LinkedHashMap<>();
+                                additionalInfo.put("installLocation", installLocation);
+                                additionalInfo.put("installSource", installSource);
+
+                                ApplicationInfo app = new ApplicationInfo(name, version, publisher, installDate,
+                                        additionalInfo);
+                                appInfoSet.add(app);
+                            } catch (Win32Exception e) {
+                                // Skip keys that are inaccessible or have missing values
                             }
-                            String version = getRegistryValueOrUnknown(rootKey, fullPath, "DisplayVersion", accessFlag);
-                            String publisher = getRegistryValueOrUnknown(rootKey, fullPath, "Publisher", accessFlag);
-                            String installDate = getRegistryValueOrUnknown(rootKey, fullPath, "InstallDate",
-                                    accessFlag);
-                            String installLocation = getRegistryValueOrUnknown(rootKey, fullPath, "InstallLocation",
-                                    accessFlag);
-                            String installSource = getRegistryValueOrUnknown(rootKey, fullPath, "InstallSource",
-                                    accessFlag);
-
-                            long installDateEpoch = ParseUtil.parseDateToEpoch(installDate, "yyyyMMdd");
-
-                            Map<String, String> additionalInfo = new LinkedHashMap<>();
-                            additionalInfo.put("installLocation", installLocation);
-                            additionalInfo.put("installSource", installSource);
-
-                            ApplicationInfo app = new ApplicationInfo(name, version, publisher, installDateEpoch,
-                                    additionalInfo);
-                            appInfoSet.add(app);
-                        } catch (Win32Exception e) {
-                            // Skip keys that are inaccessible or have missing values
                         }
+                    } catch (Win32Exception e) {
+                        // Skip paths that are inaccessible
                     }
                 }
             }
@@ -87,15 +113,114 @@ public final class InstalledAppsData {
         return new ArrayList<>(appInfoSet);
     }
 
-    private static String getRegistryValueOrUnknown(WinReg.HKEY rootKey, String path, String key, int accessFlag) {
+    private static String registryValueToString(Object registryValueOrNull) {
+        if (registryValueOrNull instanceof Integer) {
+            return Integer.toString((int) registryValueOrNull);
+        }
+        return (String) registryValueOrNull;
+    }
+
+    private static long registryValueToLong(Object registryValueOrNull) {
+        if (registryValueOrNull == null) {
+            return 0L;
+        }
+
+        // Calculate reasonable timestamp bounds (current time to 30 years ago)
+        long currentTimeSecs = System.currentTimeMillis() / 1000L;
+        long minSaneTimestamp = currentTimeSecs - THIRTY_YEARS_IN_SECS;
+        if (registryValueOrNull instanceof Integer) {
+            int value = (Integer) registryValueOrNull;
+            if (value > minSaneTimestamp && value < currentTimeSecs) {
+                return value * 1000L;
+            }
+            return value;
+        } else if (registryValueOrNull instanceof String) {
+            String dateStr = ((String) registryValueOrNull).trim();
+            // Try yyyyMMdd first
+            long epoch = ParseUtil.parseDateToEpoch(dateStr, "yyyyMMdd");
+            if (epoch == 0) {
+                // If that fails, try MM/dd/yyyy
+                epoch = ParseUtil.parseDateToEpoch(dateStr, "MM/dd/yyyy");
+            }
+            return epoch;
+        }
+        return 0L;
+    }
+
+    private static Object getRegistryValueOrNull(HKEY rootKey, String path, String key, int accessFlag) {
+        HKEY hKey = null;
         try {
-            String value = Advapi32Util.registryGetStringValue(rootKey, path, key, accessFlag);
-            if (value != null && !value.trim().isEmpty()) {
+            hKey = getRegistryHKey(rootKey, path, accessFlag);
+            Object value = registryGetValue(hKey, key);
+            if ((value instanceof Integer) || (value instanceof String && !((String) value).trim().isEmpty())) {
                 return value;
             }
         } catch (Win32Exception e) {
             LOG.trace("Unable to access " + path + " with flag " + accessFlag + ": " + e.getMessage());
+        } finally {
+            if (hKey != null) {
+                int rc = ADV.RegCloseKey(hKey);
+                if (rc != ERROR_SUCCESS) {
+                    throw new Win32Exception(rc);
+                }
+            }
         }
         return null;
+    }
+
+    private static HKEY getRegistryHKey(HKEY rootKey, String path, int accessFlag) {
+        HKEYByReference phkKey = new HKEYByReference();
+        int rc = ADV.RegOpenKeyEx(rootKey, path, 0, KEY_READ | accessFlag, phkKey);
+        if (rc != ERROR_SUCCESS) {
+            throw new Win32Exception(rc);
+        }
+        return phkKey.getValue();
+    }
+
+    private static Object registryGetValue(HKEY hKey, String value) {
+        IntByReference lpcbData = new IntByReference();
+        IntByReference lpType = new IntByReference();
+        int rc = ADV.RegQueryValueEx(hKey, value, 0, lpType, (Pointer) null, lpcbData);
+        if (rc != ERROR_SUCCESS && rc != ERROR_INSUFFICIENT_BUFFER) {
+            throw new Win32Exception(rc);
+        }
+        int type = lpType.getValue();
+        switch (type) {
+        case REG_SZ:
+        case REG_EXPAND_SZ:
+            return registryGetString(hKey, value, lpType, lpcbData);
+        case REG_DWORD:
+            return registryGetDword(hKey, value, lpType, lpcbData);
+        default:
+            LOG.warn("Unsupported registry data type {} for {}", type, value);
+            return null;
+        }
+    }
+
+    private static String registryGetString(HKEY hKey, String value, IntByReference lpType, IntByReference lpcbData) {
+        if (lpcbData.getValue() == 0) {
+            return "";
+        }
+        // Add space for wide string null terminator
+        Memory mem = new Memory(lpcbData.getValue() + Native.WCHAR_SIZE);
+        mem.clear();
+        int rc = ADV.RegQueryValueEx(hKey, value, 0, lpType, mem, lpcbData);
+        if (rc != ERROR_SUCCESS && rc != ERROR_INSUFFICIENT_BUFFER) {
+            throw new Win32Exception(rc);
+        }
+        if (W32APITypeMapper.DEFAULT == W32APITypeMapper.UNICODE) {
+            return mem.getWideString(0);
+        } else {
+            return mem.getString(0);
+        }
+    }
+
+    private static int registryGetDword(HKEY hKey, String value, IntByReference lpType, IntByReference lpcbData) {
+        IntByReference pData = new IntByReference();
+        int rc = ADV.RegQueryValueEx(hKey, value, 0, lpType, pData, lpcbData);
+        if (rc != ERROR_SUCCESS && rc != ERROR_INSUFFICIENT_BUFFER) {
+            throw new Win32Exception(rc);
+        }
+        return pData.getValue();
     }
 }

--- a/oshi-core/src/test/java/oshi/SystemInfoTest.java
+++ b/oshi-core/src/test/java/oshi/SystemInfoTest.java
@@ -43,6 +43,7 @@ import oshi.hardware.Sensors;
 import oshi.hardware.SoundCard;
 import oshi.hardware.UsbDevice;
 import oshi.hardware.VirtualMemory;
+import oshi.software.os.ApplicationInfo;
 import oshi.software.os.FileSystem;
 import oshi.software.os.InternetProtocolStats;
 import oshi.software.os.NetworkParams;
@@ -90,6 +91,9 @@ public class SystemInfoTest { // NOSONAR squid:S5786
         OperatingSystem os = si.getOperatingSystem();
 
         printOperatingSystem(os);
+
+        logger.info("Checking Installed Apps...");
+        printInstalledApps(os.getInstalledApplications());
 
         logger.info("Checking computer system...");
         printComputerSystem(hal.getComputerSystem());
@@ -163,6 +167,13 @@ public class SystemInfoTest { // NOSONAR squid:S5786
         oshi.add("Sessions:");
         for (OSSession s : os.getSessions()) {
             oshi.add(" " + s.toString());
+        }
+    }
+
+    private static void printInstalledApps(List<ApplicationInfo> installedApplications) {
+        oshi.add("Apps: ");
+        for (ApplicationInfo app : installedApplications.subList(0, 5)) {
+            oshi.add(" " + app.toString());
         }
     }
 

